daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala

```diff
@@ -254,25 +240,63 @@ object SExpr {
   /** Pattern match. */
   final case class SECase(scrut: SExpr, alts: Array[SCaseAlt]) extends SExpr with SomeArrayEquals {
     def execute(machine: Machine): Unit = {
-      machine.pushKont(KMatch(alts, machine.frame, machine.actuals, machine.env.size))
-      machine.ctrl = scrut
+      // All SEcase are converted to SECaseAtomic in Anf.flattenToAnf
+      throw SErrorCrash(s"execute: unexpected SEcase")
     }
-
     override def toString: String = s"SECase($scrut, ${alts.mkString("[", ",", "]")})"
   }
 
-  object SECase {
+  final case class SECaseAtomic(scrut: SExprAtomic, alts: Array[SCaseAlt])
+      extends SExpr
+      with SomeArrayEquals {
+    def execute(machine: Machine): Unit = {
+      val vscrut = scrut.lookupValue(machine)
+      executeMatchAlts(machine, alts, vscrut)
+    }
+  }
+
+  /** A let-expression with a single RHS */
+  final case class SELet1General(rhs: SExpr, body: SExpr) extends SExpr with SomeArrayEquals {
```

Comment:
The old implementation batched nested `let` bindings together. This saved us a trip through the main loop for each binding after the first one. What's the reason for not doing this anymore?

Reply:
Because we need a special case for `SELet1Builtin`

Reply:
Is that stopping us from batching in the not so special cases?

Reply:
it would make the code way more complicated

Reply:
I don't see how it would make the code significantly more complicated. A smart constructor for `SELet1General` could solve the problem quite easily, couldn't it?

Reply:
I will take a look...

Reply:
So if I understand, you want `SELetBuiltin` to be a smart constructor which would combine separate lets into a multi-let, constructing the expression form: `SELet(bounds: Array[SExpr], body: SExpr)`.

The trouble with this is that the special case of `SEAppAtomicSaturatedBuiltin` would be detected at run time instead of compile time as it is now.

But worse: the execution of  `SELet1General` and `SELet1Builtin` are very different - one pushes continuations, the other doesn't - one sets `machine.control` to `rhs`, the other to `body`.

I'm having trouble seeing this two different ways to execute a single-let can be easily reconciled/combined as the loop-body code for a multi `SELet` execution. I think it would involve constructing new `SELet1` expressions with suffixes of the rhss-list.

Reply:
I only meant batching `SELet1General` and leaving `SELet1Builtin` as is. Would that work? Either way, let's leave it for a separate PR in order to move forward with this.

Reply:
Can you please make a ticket for this.

Reply:
I think this ticket is still missing.

Reply:
https://github.com/digital-asset/daml/issues/6557